lib/wrappers/openssl.nim

```diff
@@ -647,3 +664,54 @@ proc md5_Str*(str: string): string =
 
 when defined(nimHasStyleChecks):
   {.pop.}
+
+# Certificates store
+
+type PX509_STORE* = SslPtr
+type PX509_OBJECT* = SslPtr
+
+{.push callconv:cdecl, dynlib:DLLUtilName, importc.}
+
+proc X509_OBJECT_new*(): PX509_OBJECT
+proc X509_OBJECT_free*(a: PX509_OBJECT)
+
+proc X509_STORE_new*(): PX509_STORE
+proc X509_STORE_free*(v: PX509_STORE)
+proc X509_STORE_lock*(ctx: PX509_STORE): cint
+proc X509_STORE_unlock*(ctx: PX509_STORE): cint
+proc X509_STORE_up_ref*(v: PX509_STORE): cint
+proc X509_STORE_set_flags*(ctx: PX509_STORE; flags: culong): cint
+proc X509_STORE_set_purpose*(ctx: PX509_STORE; purpose: cint): cint
+proc X509_STORE_set_trust*(ctx: PX509_STORE; trust: cint): cint
+proc X509_STORE_add_cert*(ctx: PX509_STORE; x: PX509): cint
+
+proc d2i_X509*(px: ptr PX509, i: ptr ptr cuchar, len: cint): PX509
+
+proc i2d_X509*(cert: PX509; o: ptr ptr cuchar): cint
+
+{.pop.}
+
+proc d2i_X509*(b: string): PX509 =
+  ## decode DER/BER bytestring into X.509 certificate struct
+  var bb = b.cstring
+  let i = cast[ptr ptr cuchar](addr bb)
+  let ret = d2i_X509(addr result, i, b.len.cint)
+  if ret.isNil:
+    raise newException(Exception, "X.509 certificate decoding failed")
+
+proc i2d_X509*(cert: PX509): string =
+  ## encode `cert` to DER string
+  let encoded_length = i2d_X509(cert, nil)
+  result = newString(encoded_length)
+  var q = result.cstring
+  let o = cast[ptr ptr cuchar](addr q)
+  let length = i2d_X509(cert, o)
+  if length.int <= 0:
+    raise newException(Exception, "X.509 certificate encoding failed")
+
+when isMainModule:
+  let certbytes = readFile("certificate.der")
+  let cert = d2i_X509(certbytes)
+  let encoded = cert.i2d_X509()
+  assert encoded == certbytes
+  echo "OK"
```

Comment:
remove that line?

Reply:
That is a test..

Reply:
even if it's a test, it's not needed; see for example isMainModule tests inside lib/pure/math.nim
`logging should be reserved for the exceptional, not the expected` is a good rule of thumb